File: go/webhooks.go
package legalize

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"math"
	"strconv"
	"strings"
	"time"
)

// DefaultWebhookTolerance is the default allowable clock drift
// between the signed timestamp and the verifier's wall clock. Five
// minutes matches the server defaults and the Python / Node SDKs.
const DefaultWebhookTolerance = 5 * time.Minute

// supportedSchemes lists the signature-version prefixes the SDK
// accepts today. Future schemes will be added here; unknown prefixes
// are silently ignored per the Stripe-style scheme.
var supportedSchemes = map[string]struct{}{"v1": {}}

// WebhookEvent is the parsed payload of a verified webhook delivery.
type WebhookEvent struct {
	ID        string          `json:"id"`
	Type      string          `json:"type"`
	CreatedAt string          `json:"created_at"`
	Data      json.RawMessage `json:"data"`
	Raw       json.RawMessage `json:"-"`
}

// VerifyOption tunes Verify's behaviour.
type VerifyOption func(*verifyConfig)

type verifyConfig struct {
	tolerance time.Duration
	now       time.Time
}

// WithTolerance overrides the default ±5 minute clock drift window.
func WithTolerance(d time.Duration) VerifyOption {
	return func(c *verifyConfig) { c.tolerance = d }
}

// WithReferenceTime lets tests inject a deterministic wall clock.
// Real integrations should leave this unset.
func WithReferenceTime(t time.Time) VerifyOption {
	return func(c *verifyConfig) { c.now = t }
}

// Verify checks a webhook signature and returns the parsed event.
//
// Inputs are the raw request body (not a re-serialised JSON), the
// X-Legalize-Signature header, the X-Legalize-Timestamp header, and
// the endpoint's signing secret.
//
// On failure it returns a *WebhookVerificationError whose Reason is
// one of: "missing_header", "bad_timestamp", "timestamp_outside_tolerance",
// "no_valid_signature", "bad_signature".
//
// The signature is compared with crypto/subtle.ConstantTimeCompare.
// Verification happens BEFORE the JSON body is parsed, protecting
// against JSON-parser resource exhaustion on unauthenticated bodies.
func Verify(payload []byte, sigHeader, timestamp, secret string, opts ...VerifyOption) (*WebhookEvent, error) {
	cfg := verifyConfig{tolerance: DefaultWebhookTolerance}
	for _, opt := range opts {
		opt(&cfg)
	}

	if sigHeader == "" || timestamp == "" || secret == "" {
		return nil, &WebhookVerificationError{Reason: "missing_header"}
	}

	// Timestamp parse.
	tsInt, err := strconv.ParseInt(strings.TrimSpace(timestamp), 10, 64)
	if err != nil {
		return nil, &WebhookVerificationError{Reason: "bad_timestamp"}
	}

	reference := cfg.now
	if reference.IsZero() {
		reference = time.Now()
	}
	diff := reference.Unix() - tsInt
	if math.Abs(float64(diff)) > cfg.tolerance.Seconds() {
		return nil, &WebhookVerificationError{Reason: "timestamp_outside_tolerance"}
	}

	// Compute expected signature.
	expectedHex := ComputeSignatureHex(secret, payload, timestamp)
	expectedBytes, err := hex.DecodeString(expectedHex)
	if err != nil {
		// Shouldn't happen — we control the producer.
		return nil, &WebhookVerificationError{Reason: "bad_signature"}
	}

	candidates := extractSchemeHexes(sigHeader)
	if len(candidates) == 0 {
		return nil, &WebhookVerificationError{Reason: "no_valid_signature"}
	}

	matched := false
	for _, cand := range candidates {
		candBytes, err := hex.DecodeString(cand)
		if err != nil {
			continue
		}
		if subtle.ConstantTimeCompare(candBytes, expectedBytes) == 1 {
			matched = true
			break
		}
	}
	if !matched {
		return nil, &WebhookVerificationError{Reason: "bad_signature"}
	}

	// Parse JSON payload AFTER verification succeeds.
	var raw map[string]any
	if err := json.Unmarshal(payload, &raw); err != nil {
		return nil, &WebhookVerificationError{Reason: "bad_signature"}
	}

	evt := &WebhookEvent{Raw: payload}
	if v, ok := raw["id"].(string); ok {
		evt.ID = v
	}
	if v, ok := raw["event_type"].(string); ok {
		evt.Type = v
	} else if v, ok := raw["type"].(string); ok {
		evt.Type = v
	}
	if v, ok := raw["created_at"].(string); ok {
		evt.CreatedAt = v
	}
	if d, ok := raw["data"]; ok {
		if b, err := json.Marshal(d); err == nil {
			evt.Data = b
		}
	}
	return evt, nil
}

// ComputeSignatureHex returns the canonical hex signature for the
// given (secret, payload, timestamp). Exposed so tests can build
// known-good signature vectors and callers can implement custom flows.
func ComputeSignatureHex(secret string, payload []byte, timestamp string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(timestamp))
	mac.Write([]byte("."))
	mac.Write(payload)
	return hex.EncodeToString(mac.Sum(nil))
}

// ComputeSignature returns the "v1=<hex>" header-ready signature.
func ComputeSignature(secret string, payload []byte, timestamp string) string {
	return "v1=" + ComputeSignatureHex(secret, payload, timestamp)
}

// extractSchemeHexes parses an X-Legalize-Signature header for every
// supported "vN=<hex>" pair. Unknown schemes are ignored; a malformed
// or empty header yields no candidates (and the caller MUST treat
// that as a verification failure).
func extractSchemeHexes(header string) []string {
	var out []string
	for _, part := range strings.Split(header, ",") {
		part = strings.TrimSpace(part)
		scheme, value, ok := strings.Cut(part, "=")
		if !ok {
			continue
		}
		if _, ok := supportedSchemes[scheme]; !ok {
			continue
		}
		value = strings.TrimSpace(value)
		if value == "" {
			continue
		}
		out = append(out, value)
	}
	return out
}

File: node/src/webhooks.ts
/**
 * Webhook signature verification.
 *
 * Mirrors the server-side signer in the Legalize web backend. The scheme
 * is Stripe-shaped:
 *
 *   - Signed content: `${timestamp}.${rawJsonBody}` — RAW bytes, never
 *     a re-serialized JSON string. Reserialization will break the signature.
 *   - Algorithm: HMAC-SHA256, hex-encoded.
 *   - Header format: `v1=<hex>`. A future v2 scheme can coexist:
 *     `X-Legalize-Signature: v1=<sig1>,v2=<sig2>`.
 *   - Replay protection: reject if the header timestamp is more than
 *     `tolerance` seconds away from `now`. Default 300 seconds.
 *
 * Usage (Express):
 *
 *   import express from "express";
 *   import { Webhook, WebhookVerificationError } from "@legalize-dev/sdk";
 *
 *   app.post("/webhooks/legalize",
 *     express.raw({ type: "application/json" }),
 *     (req, res) => {
 *       try {
 *         const event = Webhook.verify({
 *           payload: req.body,  // Buffer
 *           sigHeader: req.header("X-Legalize-Signature") ?? "",
 *           timestamp: req.header("X-Legalize-Timestamp") ?? "",
 *           secret: process.env.LEGALIZE_WHSEC!,
 *         });
 *         if (event.type === "law.updated") { ... }
 *         res.status(204).send();
 *       } catch (err) {
 *         res.status(400).send();
 *       }
 *     });
 */

import { createHmac, timingSafeEqual } from "node:crypto";

import { WebhookVerificationError } from "./errors.js";

export const DEFAULT_TOLERANCE_SECONDS = 300; // 5 minutes — matches server
const SUPPORTED_SCHEMES = ["v1"];

export interface WebhookEvent {
  /** Server-assigned event id (e.g. `evt_...`). */
  readonly id: string;
  /** Event type (`law.created`, `law.updated`, `law.repealed`, ...). */
  readonly type: string;
  /** Server-side timestamp (ISO-8601 string). */
  readonly createdAt: string;
  /** Event-specific payload body. */
  readonly data: Record<string, unknown>;
  /** The full decoded JSON body — useful for fields the SDK doesn't type. */
  readonly raw: Record<string, unknown>;
}

export interface WebhookVerifyOptions {
  /** The raw request body bytes. Do NOT pass a re-serialized JSON string. */
  payload: Buffer | Uint8Array | string;
  /** The `X-Legalize-Signature` header value. May contain several `vN=<hex>` pairs. */
  sigHeader: string;
  /** The `X-Legalize-Timestamp` header value (Unix seconds, decimal string). */
  timestamp: string;
  /** The endpoint's signing secret. */
  secret: string;
  /** Seconds of clock skew accepted. Default 300. */
  tolerance?: number;
  /** Unit-test hook: override the reference wall clock (Unix seconds). */
  now?: number;
}

export class Webhook {
  /** Default anti-replay tolerance. */
  static readonly TOLERANCE = DEFAULT_TOLERANCE_SECONDS;

  /** Compute the canonical `v1=<hex>` signature for (payload, timestamp). */
  static computeSignature(secret: string, payload: Buffer | Uint8Array | string, timestamp: string): string {
    const payloadBuf = normalizePayload(payload);
    const signed = Buffer.concat([
      Buffer.from(timestamp, "utf8"),
      Buffer.from(".", "utf8"),
      payloadBuf,
    ]);
    const sig = createHmac("sha256", secret).update(signed).digest("hex");
    return `v1=${sig}`;
  }

  /**
   * Verify a webhook delivery and return the parsed event.
   *
   * Signature verification happens BEFORE JSON parsing, to protect the
   * process from resource-exhaustion on unauthenticated bodies.
   *
   * Throws WebhookVerificationError on any failure. The `.reason`
   * field identifies which check tripped for server-side logging,
   * while the user-facing message stays generic.
   */
  static verify(options: WebhookVerifyOptions): WebhookEvent {
    if (!options.sigHeader || !options.timestamp || !options.secret) {
      throw new WebhookVerificationError("missing_header");
    }

    const payloadBuf = normalizePayload(options.payload);

    // ---- timestamp (anti-replay) --------------------------------------
    if (!/^-?\d+$/.test(options.timestamp.trim())) {
      throw new WebhookVerificationError("bad_timestamp");
    }
    const ts = parseInt(options.timestamp.trim(), 10);
    if (Number.isNaN(ts)) {
      throw new WebhookVerificationError("bad_timestamp");
    }
    const tol = options.tolerance ?? Webhook.TOLERANCE;
    const reference = options.now ?? Date.now() / 1000;
    if (Math.abs(reference - ts) > tol) {
      throw new WebhookVerificationError("timestamp_outside_tolerance");
    }

    // ---- signature ----------------------------------------------------
    const expected = Webhook.computeSignature(options.secret, payloadBuf, options.timestamp);
    const expectedHex = expected.slice(expected.indexOf("=") + 1);
    const candidates = extractSchemeHexes(options.sigHeader);
    if (candidates.length === 0) {
      throw new WebhookVerificationError("no_valid_signature");
    }
    let match = false;
    const expectedBuf = Buffer.from(expectedHex, "hex");
    for (const candidate of candidates) {
      let candBuf: Buffer;
      try {
        candBuf = Buffer.from(candidate, "hex");
      } catch {
        continue;
      }
      if (candBuf.length !== expectedBuf.length) continue;
      if (timingSafeEqual(expectedBuf, candBuf)) {
        match = true;
        break;
      }
    }
    if (!match) {
      throw new WebhookVerificationError("bad_signature");
    }

    // ---- parse --------------------------------------------------------
    let parsed: unknown;
    try {
      parsed = JSON.parse(payloadBuf.toString("utf8"));
    } catch {
      throw new WebhookVerificationError("bad_signature");
    }
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      throw new WebhookVerificationError("bad_signature");
    }
    return webhookEventFromPayload(parsed as Record<string, unknown>);
  }
}

function normalizePayload(payload: Buffer | Uint8Array | string): Buffer {
  if (Buffer.isBuffer(payload)) return payload;
  if (typeof payload === "string") return Buffer.from(payload, "utf8");
  return Buffer.from(payload);
}

function extractSchemeHexes(header: string): string[] {
  const out: string[] = [];
  for (const rawPart of header.split(",")) {
    const part = rawPart.trim();
    const eq = part.indexOf("=");
    if (eq < 0) continue;
    const scheme = part.slice(0, eq);
    const value = part.slice(eq + 1).trim();
    if (!value) continue;
    if (!SUPPORTED_SCHEMES.includes(scheme)) continue;
    // Hex validity is checked later via Buffer.from + length match.
    if (!/^[0-9a-fA-F]+$/.test(value)) continue;
    out.push(value);
  }
  return out;
}

function webhookEventFromPayload(payload: Record<string, unknown>): WebhookEvent {
  const eventType =
    (payload.event_type as string | undefined) ??
    (payload.type as string | undefined) ??
    "";
  return {
    id: String(payload.id ?? ""),
    type: String(eventType),
    createdAt: String(payload.created_at ?? ""),
    data:
      payload.data && typeof payload.data === "object" && !Array.isArray(payload.data)
        ? { ...(payload.data as Record<string, unknown>) }
        : {},
    raw: { ...payload },
  };
}
